adds key management for multiple versioned keys

alecps opened the pull request:
### Description

Adds support for multiple versioned keys to the ODIS signer's key management code and implements the following key rotation protocol. 

1. Signers add the new key to their key vault
2. The signers should update the "latest" version of the key in their env file so that this configuration goes into effect and the key is prefetched the next time they redeploy their signer.
3. The new keys must be named according to the patterns `${env.PHONE_NUMBER_PRIVACY_KEY_NAME_BASE}-{env.PHONE_NUMBER_PRIVACY_LATEST_KEY_VERSION}` or `${env.DOMAINS_KEY_NAME_BASE}-{env.DOMAINS_LATEST_KEY_VERSION}`. Note that these names are configurable via the .env file, but must be suffixed by a '-' followed by the latest key version number (key versions start at 1). If base key names are not configured, they default to 'phoneNumberPrivacy' and 'domains' respectively (ex. 'phoneNumberPrivacy-1' would be the default name for the original ODIS Key). 
4. The combiner's config file is updated with the new latest version of the key only once all signers have confirmed that they have correctly stored the new key in their keystore.
5. The combiner is redeployed, and will now send requests to signers with an incremented 'keyVersion' header. Upon receiving requests for this new keyVersion, signers will fetch it from their keyvaults and begin using it to sign requests. 
6. The combiner monitors signer responses and ensures they are signing with the correct keyVersion
7. Once the signers have confirmation that the system is operating properly and the key rotation went smoothly, they can delete their old key shares.

### Other changes

None

### Tested

Key rotation will be tested manually in staging and alfajores as part of CIP-40 release process and e2e testing 

### Related issues

- Fixes #8660 

### Backwards compatibility

Yes

### Documentation

Added to README

nategraf commented on packages/phone-number-privacy/combiner/src/config.ts:
```diff
@@ -88,6 +96,10 @@ if (DEV_MODE) {
     },
+    keyVersions: {
+      phoneNumberPrivacy: functionConfig.keyVersions.phoneNumberPrivacy || 1,
+      domains: functionConfig.keyVersions.domains || 1,
```
Using `??` instead of `||` is more ideomatic.
```suggestion
      phoneNumberPrivacy: functionConfig.keyVersions.phoneNumberPrivacy ?? 1,
      domains: functionConfig.keyVersions.domains ?? 1,
```

nategraf commented on packages/phone-number-privacy/combiner/src/signing/get-threshold-signature.ts:
```diff
@@ -73,6 +73,8 @@ async function requestSignatures(request: Request, response: Response) {
   })
   obs.observe({ entryTypes: ['measure'], buffered: true })
 
+  request.body.keyVersion = config.keyVersions.phoneNumberPrivacy
```
It doesn't seem quite right to me to mutate the request object. In addition to a general avoidance of mutating function inputs, the client may have specified `keyVersion`, in which case we should honor it if possible.

nategraf commented on packages/phone-number-privacy/signer/src/domain/domain.service.ts:
```diff
@@ -139,6 +140,11 @@ export class DomainService implements IDomainService {
 
+    const key: Key = {
+      type: KeyType.domains,
+      version: 1,
```
We should add this to the signature request object.

nategraf replied:
Just realized that this is a challenge because we need to be able to have the combiner decide which key the request uses, if the client does not specify. My inclination is to either A) have the request be placed in a `body.request` field, and add a `body.combiner` field to allow for the combiner to transmit metadata such as this or B) add a custom http header. B is slightly less elegant, but it does have the advantage of working for both the phone number privacy and the domains API.

nategraf replied:
Maybe you can think of a better solution.

alecps replied:
I ended up going with option B. Let me know what you think

nategraf commented on packages/phone-number-privacy/signer/src/key-management/azure-key-provider.ts:
```diff
@@ -1,16 +1,16 @@
       const keyVaultClient = new AzureKeyVaultClient(vaultName)
-      const privateKey = await keyVaultClient.getSecret(secretName)
-      this.setPrivateKey(privateKey)
+      const privateKey = await keyVaultClient.getSecret(`${key.type}${key.version}`)
```
Can we insert a delimiter like a dash or slash to separate the version from the type (e.g. `domains-1`, or `domains/1`). Not super important, but would be nice.

nategraf commented on packages/phone-number-privacy/signer/src/key-management/key-provider-base.ts:
```diff
@@ -1,28 +1,51 @@
+  public getPrivateKey(key: Key) {
+    const privateKey = this.privateKeys.get(key)
+    if (!privateKey) {
       throw new Error('Private key is empty, provider not properly initialized')
```
It's now the case that the user could have requested a key version that does not exist. Suggested change to the error message accounts for that.
```suggestion
      throw new Error(`Private key is unavailable: ${JSON.stringify(key)}`)
```

nategraf commented on packages/phone-number-privacy/signer/src/key-management/key-provider-base.ts:
```diff
@@ -1,28 +1,51 @@
+    type: KeyType.domains,
+    version: 1,
+  },
+]
```
Does this mean that we will need to modify this code and deploy a new version of the signer if we do a key resharing? If so, it would be better if this could be controlled through a config value.

nategraf replied:
It's also unclear to me where this is used. Because keys can be loaded on demand, it doesn't seem like we need to have a list of all key versions in the code itself (e.g. instead we could attempt to fetch a key we a user request comes in that requires it, although I can see the advantages of prefetching).

alecps replied:
I ended up doing a compromise where we prefetch the "latest" version of each key but also fetch keys dynamically if the combiner requests them

nategraf commented on packages/phone-number-privacy/signer/src/signing/get-partial-signature.ts:
```diff
@@ -43,6 +45,13 @@ export async function handleGetBlindedMessagePartialSig(
   logger.info({ request: request.body }, 'Request received')
   logger.debug('Begin handleGetBlindedMessagePartialSig')
 
+  const keyVersion = request.body.keyVersion ?? 1
```
Default should probably be the latest version instead of version 1. We will want to delete key version 1 soon for security reasons.

nategraf commented on packages/phone-number-privacy/signer/src/signing/get-partial-signature.ts:
```diff
@@ -43,6 +45,13 @@ export async function handleGetBlindedMessagePartialSig(
   logger.debug('Begin handleGetBlindedMessagePartialSig')
 
+  const keyVersion = request.body.keyVersion ?? 1
+  delete request.body.keyVersion
```
Why delete the value from the request object? In general we shouldn't mutate the request object, and this especially seems odd to me.

nategraf replied:
Oh, is tis to make sure the signature check will pass? Hmmm, ya this is a bit of a challenge.

nategraf commented on packages/phone-number-privacy/signer/src/key-management/key-provider.ts:
```diff
@@ -26,8 +26,8 @@ export async function initKeyProvider() {
   }
-  logger.info('Fetching key')
-  await keyProvider.fetchPrivateKeyFromStore()
+  logger.info('Fetching keys')
```
```suggestion
  logger.info(`Fetching keys: ${JSON.stringify(keys)}`)
```

nategraf commented:
Nice work overall.

One disadvantage of this approach is that we are now requiring operators to name their keys in a specific way wheras before they could use any name they like. This is probably fine, but it would be nice to allow them to override the default key paths (derived from the key `Key` obejct) with a custom path in the config file.

The code itself also does not seem to support deploying a new key version (i.e. executing resharing) without code changes and redeployment of the signer. Ideally it would only require config changes, or even only placing the new key in the key store. This will be important if we want to start running key resharing proactively. We should also keep in mind that the goal is to delete the old key shares, which we would do after everyone has upgraded to the new shares.

Commit:
combiner changes

Commit:
(drive-by) fixes matchmaking fullnode error handling

Commit:
adds default key names to common package

Commit:
allow for key names and latest versions to be configured via env file

Commit:
allow for keys to be fetched dynamically

Commit:
prefetch latest version of each key

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into alecps/odisKeyRotation

Commit:
log keyVersion in combiner when signer responds

alecps commented:
> Nice work overall.
> 
> One disadvantage of this approach is that we are now requiring operators to name their keys in a specific way wheras before they could use any name they like. This is probably fine, but it would be nice to allow them to override the default key paths (derived from the key `Key` obejct) with a custom path in the config file.
> 
> The code itself also does not seem to support deploying a new key version (i.e. executing resharing) without code changes and redeployment of the signer. Ideally it would only require config changes, or even only placing the new key in the key store. This will be important if we want to start running key resharing proactively. We should also keep in mind that the goal is to delete the old key shares, which we would do after everyone has upgraded to the new shares.

Agreed. Let me know if this updated implementation addresses your concerns

Commit:
adds key rotation documentation to README

Commit:
fixes error

Commit:
fixes key management tests

Commit:
fixes lint error

Commit:
fixes tests again

Commit:
fixes tests again again

nategraf commented on packages/phone-number-privacy/combiner/src/signing/get-threshold-signature.ts:
```diff
@@ -73,6 +73,8 @@ async function requestSignatures(request: Request, response: Response) {
   })
   obs.observe({ entryTypes: ['measure'], buffered: true })
 
+  request.headers.keyVersion = config.keyVersions.phoneNumberPrivacy.toString()
```
Convention for HTTP(S) headers is to name them in Hyphen-Delimited-Camel-Case. Additionally, it's important to be reasonably confident that the name will not collide with a header provided by any other application.
```suggestion
  request.headers['Odis-Key-Version'] = config.keyVersions.phoneNumberPrivacy.toString()
```

nategraf commented on packages/phone-number-privacy/combiner/src/signing/get-threshold-signature.ts:
```diff
@@ -182,6 +184,8 @@ async function handleSuccessResponse(
   controller: AbortController
 ) {
   const logger: Logger = response.locals.logger
+  const keyVersion: number = Number(response.header('keyVersion'))
```
Based on the suggestion above.
```suggestion
  const keyVersion: number = Number(response.header('Odis-Key-Version'))
```

nategraf commented on packages/phone-number-privacy/common/src/interfaces/requests.ts:
```diff
@@ -15,6 +15,11 @@ import {
+  phoneNumberPrivacy = 'phoneNumberPrivacy',
+  domains = 'domains',
+}
+
```
IIUC, these are the defaults used by the signer. Since it is only used in that package, and the client will not know what name the signer uses for their keys, I'd recommend this be moved into the signer package.

On a less important note, I am trying to make our enum conventions more consistent. We use CapitalizedCamelCase and CAPITALIZED_SNAKE_CASE in different places, but I think the latter is the wider used standard.
```suggestion
export enum DefaultKeyName {
  PHONE_NUMBER_PRIVACY = 'phoneNumberPrivacy',
  DOMAINS = 'domains',
}
```

nategraf commented on packages/phone-number-privacy/signer/README.md:
```diff
@@ -99,6 +106,17 @@ Then check on the service to make sure its running:
+3. The combiner operator will update the combiner to request the new key share version via a custom request header field once all signers are ready.
+4. The signers will fetch the new key shares from their keystores upon receiving these requests.
+5. When the combiner operator sees that all signers are signing with the new key share and confirms that the system is healthy, signers will be instructed to delete their old key shares.
+6. After deleting their old key shares, signers should increment `PHONE_NUMBER_PRIVACY_LATEST_KEY_VERSION` or `DOMAINS_LATEST_KEY_VERSION` accordingly. This will instruct the signer to prefetch this key version the next time it starts up.
```
IIUC, the signers should do this before deleting the old key shares, and may do it immediately upon installing the new key. If so, this would be either step 5 (before deleting) or step 2 (after adding the new key) respectively. Is that accurate?

alecps replied:
Good point. It doesn't make a big difference when the signer updates these variables since prefetching is really just an optimization. I'll make it step 2

nategraf commented on packages/phone-number-privacy/signer/src/domain/domain.service.ts:
```diff
@@ -139,6 +141,11 @@ export class DomainService implements IDomainService {
 
+    const key: Key = {
+      name: KeyName.domains,
+      version: Number(request.headers.keyVersion) || config.keystore.keys.domains.latest,
```
Suggestions to adjust the header name and to use `??` instead of `||`, which prevents the case of `0` from being handled as `undefined`. I'm also not cure what happens with `NaN` in the current code.
```suggestion
      version: Number(request.headers['Odis-Key-Version'] ?? config.keystore.keys.domains.latest),
```

nategraf commented on packages/phone-number-privacy/signer/src/key-management/key-provider-base.ts:
```diff
@@ -1,28 +1,58 @@
 
-    return this.privateKey
+  public async getPrivateKeyOrFetchFromStore(key: Key): Promise<string> {
+    if (key.version < 1 || key.version > 10) {
```
Unless there is a particular reason for the restriction, I would not include it. If we implement proactive resharing, we would see more than 10 key versions. Without it, requesting a key version too high seems harmless to me, but I could be missing something.
```suggestion
    if (key.version < 1) {
```

nategraf commented on packages/phone-number-privacy/signer/src/key-management/key-provider-base.ts:
```diff
@@ -1,28 +1,58 @@
+      case KeyName.phoneNumberPrivacy:
+        return config.keystore.keys.phoneNumberPrivacy.name || key.name
+      case KeyName.domains:
+        return config.keystore.keys.domains.name || key.name
```
Using the `??` gives 10 points to Gryffindor
```suggestion
        return config.keystore.keys.phoneNumberPrivacy.name ?? key.name
      case KeyName.domains:
        return config.keystore.keys.domains.name ?? key.name
```

alecps replied:
I think this might also be one of the few places where || is appropriate over ??. My reasoning here is that the env variables that we use to set the config values could be empty strings, in which case we would want to override that value with key.name. In general, I think we'd want to override any falsy value here but I may be missing something.

nategraf commented on packages/phone-number-privacy/signer/src/signing/get-partial-signature.ts:
```diff
@@ -43,6 +46,11 @@ export async function handleGetBlindedMessagePartialSig(
 
+  const key: Key = {
+    name: KeyName.phoneNumberPrivacy,
+    version: Number(request.headers.keyVersion) || config.keystore.keys.phoneNumberPrivacy.latest,
```
Same suggestion as above.
```suggestion
    version: Number(request.headers['Odis-Key-Version'] ?? config.keystore.keys.phoneNumberPrivacy.latest),
```

alecps replied:
I think this might be one of the few places where || is appropriate over ??. My reasoning is the Number() could return NaN which is falsy but not null/undefined. We also don't support version '0' of keys. I'm basing this off the following stackoverflow post. Let me know if that checks out.  https://stackoverflow.com/questions/23437476/in-typescript-how-to-check-if-a-string-is-numeric/23440948#23440948

nategraf commented on packages/phone-number-privacy/signer/src/signing/get-partial-signature.ts:
```diff
@@ -186,7 +192,7 @@ export async function handleGetBlindedMessagePartialSig(
     Counters.responses.labels(Endpoints.GET_BLINDED_MESSAGE_PARTIAL_SIG, '200').inc()
     logger.info({ response: signMessageResponse }, 'Signature retrieval success')
-    response.json(signMessageResponse)
+    response.set('keyVersion', key.version.toString()).json(signMessageResponse)
```
Related to suggestions elsewhere
```suggestion
    response.set('Odis-Key-Version', key.version.toString()).json(signMessageResponse)
```

nategraf commented on packages/phone-number-privacy/combiner/src/signing/get-threshold-signature.ts:
```diff
@@ -182,6 +184,8 @@ async function handleSuccessResponse(
 ) {
   const logger: Logger = response.locals.logger
+  const keyVersion: number = Number(response.header('keyVersion'))
+  logger.info({ keyVersion }, 'Signer responded with keyVersion')
```
If the key is not the version the combiner is expecting / prepared to accept, we should probably discard the response. The risk here is that including signature into the BLS aggregation will result in the whole batch being corrupted. Even if the client can recover from it, which I am not sure we've implemented, it will waste a lot of time.

nategraf reviewed (approved):
Some minor suggestions, but looks good to me!

One last suggestion would be to add a test that the phone number privacy and domains services handle the new key version header correctly (i.e. test that they use it to select which key they should use)

Commit:
responds to subset of feedback

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into alecps/odisKeyRotation

Commit:
small edit to README

Commit:
change enum naming convention

Commit:
moves DefaultKeyName to signer package

Commit:
change enum naming convention in more places

Commit:
combiner discards incorrect key version responses

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into alecps/odisKeyRotation

Commit:
increment combiner version number

Commit:
adds TODO for public polynomial management

Commit:
adds support for multiple polynomials to combiner

Commit:
fixes signer tests

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into alecps/odisKeyRotation

Commit:
fixes error

Commit:
fixes lint error

alecps commented:
Important: These changes will break on deployment if we do not update the signer keyvault configuration ahead of time. We should merge this only after we have added the new keys (under the correct names) to the vaults.

TODO: Investigate whether these changes require full celotool redeployment

codyborn commented on packages/phone-number-privacy/signer/README.md:
```diff
@@ -124,6 +131,18 @@ Then check on the service to make sure its running:
+After a key resharing, signers should rotate their key shares as follows:
+
+1. Store the new key share in the keystore according to the naming convention specified in the [Keystores](#keystores) section above.
+2. Increment `PHONE_NUMBER_PRIVACY_LATEST_KEY_VERSION` or `DOMAINS_LATEST_KEY_VERSION` as appropriate. This will instruct the signer to prefetch this new key version the next time it starts up, but there is no need to restart the signer at this point.
```
I worry about unintended reboots or scaling up which would trigger restart of the service. WDYT about indexing the keys on polynomial and the Combiner can request a specific key from the signer. This forces the Combiner deployer to not forget the polynomial and the Combiner can easily roll back to an old version upon failure without having to coordinate with the signers.

alecps replied:
Just documenting that we discussed this in a meeting and determined that given the length of the public polynomials it would be difficult to index this way. It's worth noting also that the combiner will be able to roll back to an older version so long as the signers have not removed that key version from their vaults

Commit:
update deployment configuration files

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into alecps/odisKeyRotation

Commit:
fixes error

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into alecps/odisKeyRotation

Commit:
deactivate domain signer endpoints

Commit:
add back old key name env variables

Commit:
add old key name as a fallback

Commit:
cosmetic change

Commit:
fix error handling for fallback secretName

Commit:
add checks for empty strings from undefined env vars

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into alecps/odisKeyRotation

Commit:
Merge branch 'master' into alecps/odisKeyRotation

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into alecps/odisKeyRotation

Commit:
Merge branch 'alecps/odisKeyRotation' of github.com:celo-org/celo-monorepo into alecps/odisKeyRotation

Commit:
fixes lint error

Commit:
hopefully fixes test

Commit:
hopefully also fixes test

Commit:
update package versions

Commit:
add constructor to key-provider-base

alecps commented:
closing in favor of https://github.com/celo-org/celo-monorepo/pull/9161
